### PR TITLE
For #7698: Adds search back button animation

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt [hunk 1]
```diff
@@ -352,7 +357,17 @@ class SearchFragment : Fragment(), UserInteractionHandler {
         }
     }
 
+    private fun animateBackButtonAway() {
+        val xTranslation =
+            requireView().back_button.width.toFloat() * BACK_BUTTON_ANIMATION_WIDTH_MULTIPLIER
+
+        requireView().back_button
+            .animate()
+            .translationX(xTranslation).interpolator = FastOutSlowInInterpolator()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt]: Nit: would you mind line breaking before `.interpolator...`?

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt [hunk 2]
```diff
@@ -352,7 +357,17 @@ class SearchFragment : Fragment(), UserInteractionHandler {
         }
     }
 
+    private fun animateBackButtonAway() {
+        val xTranslation =
+            requireView().back_button.width.toFloat() * BACK_BUTTON_ANIMATION_WIDTH_MULTIPLIER
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt]: It looks like this constant is to make up for padding + margin?  That works, but it's a little bit messy, opens us up to weird bugs if we ever adjust either of those values, and requires a reader to figure out what it's handling.  What about something like this:

```kotlin
val backButton = requireView().back_button
val xTranslation = with(backButton) {
    -(width + marginStart + paddingStart).toFloat()
}
```

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt [hunk 3]
```diff
@@ -198,8 +200,11 @@ class SearchFragment : Fragment(), UserInteractionHandler {
             qrFeature.get()?.scan(R.id.container)
         }
 
-        view.back_button.setOnClickListener {
-            findNavController().navigateUp()
+        view.back_button.apply {
+            setOnClickListener {
+                animateBackButtonAway()
+                findNavController().navigateUp()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt]: I think it's a bad sign that this animates then navigates up, while the ToolbarView calls the interactor, which calls the controller, which... animates then navigates up.  This is one of the things our architecture is meant to unify, so let's clean it up while we're here.

If we:
1) move `animateBackButtonAway` off of `ToolbarView` and on to `SearchController`
2) replace 205 & 206 in `SearchFragment` with `searchInteractor.onEditingCanceled()`

then we have a more reasonable and consistent data flow.

It seems to me that we want back press and back button clicks to behave the same here, so I'm happy to have them call the same interactor method.  If that ever changes in the future, we can break them out as needed.

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt (hunk 2 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt]: Seems reasonable to me!

File: app/src/main/java/org/mozilla/fenix/search/SearchController.kt [hunk 4]
```diff
@@ -76,6 +76,7 @@ class DefaultSearchController(
     }
 
     override fun handleEditingCancelled() {
+        store.dispatch(SearchFragmentAction.EditingCancelled)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/SearchController.kt]: Nit: there's nothing wrong with this name, but just to align with AC I'd prefer to name it `UpdateEditingCanceled`.

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt [hunk 5]
```diff
@@ -352,6 +363,18 @@ class SearchFragment : Fragment(), UserInteractionHandler {
         }
     }
 
+    private fun animateBackButtonAway() {
+        val backButton = requireView().back_button
+        val xTranslation = with(backButton) {
+            -(width + marginStart + paddingStart).toFloat()
+        }
+
+        requireView().back_button
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt]: Nit: you can replace this line with `backButton`

**Pull request merged**